README.txt
URL: http://www.emacswiki.org/cgi-bin/wiki.pl?ErBot for all erbot
information, and installation help. 



The files in the `contrib' directory are optional "third-party"
add-ons that are used for erbot.  Not all of them are strictly
required.  The versions here are provided for convenience, and are
also the versions known to work with erbot, but you might want to
fetch their latest versions from their respective authors' websites.

The file `examples/dotemacs-mybot' is meant to be the bot's .emacs
file.  Copy it to .emacs in your bot's home directory, and then make
any changes you like.

NOTE: A data file for contrib/units.el may be obtained in Debian by
performing "apt-get install units", and adding
  (setq units-dat-file "/usr/share/misc/units.dat")
to the bot's .emacs.

To compile the erbot source, edit Makefile and run "make".  To install
it, edit Makefile and run "make install".


====================================================

For developers: 


Namespaces used by these files: fs-, fs.*-, erb.*-



Next, we attempt to describe the various abbreviations and terms used
in this package.


+---------+------------------------------------------------------------+
|botbbdb  |Name of the bbdb database file used by the bots.            |
+---------+------------------------------------------------------------+
|erb      |ErBot stands for erc robot, and started out from            |
|         |erc-robot.el.  We named the new file erbot, and most        |
|         |namespaces start with erb.                                  |
+---------+------------------------------------------------------------+
|         |                                                            |
+---------+------------------------------------------------------------+
|erbc-    |This referred to erbot-comands. Functions starting with this|
|         |name were availabel to public to frob as they like.  Same   |
|         |for variables.  This was replaced by fs-                    |
+---------+------------------------------------------------------------+
|erbnoc-  |This is like erbc-, except that these commands are NOT      |
|         |available to the general public (at this time).  This one is|
|         |still in use, unlike fs-.  These functions are NOT          |
|         |world-executable or world-writable, but are maintained      |
|         |alognside erbc- functions ..  erbnoc meansd: erb -          |
|         |no-commands... We have now shortened it to erbn-            |
+---------+------------------------------------------------------------+
|erbn-    |Shortening of erbnoc-                                       |
+---------+------------------------------------------------------------+
|fs-      |fsbot is a popular instance of erbot.  At some point, all   |
|         |erbc- prefixes were replaced by fs- for easier read.  Thus, |
|         |to reiterate, these functions are world-readable, weritable |
|         |writable and executable.  (The only exeptions are those that|
|         |are internally converted from fsi-, which are converted to  |
|         |fs- with a special disabled property.)  Summary: rwx for irc|
|         |users.                                                      |
+---------+------------------------------------------------------------+
|fsi-     |Like fs- but these functions and variables are only         |
|         |world-readable and world-executable, but NOT world-writable.|
|         |The "i" stands for immutable (or is it "internal"?).  These |
|         |functions are internally converted to fs- functionserbot    |
|         |usage through erbot-install-symbols.  Summary: r-x for irc  |
|         |users.                                                      |
+---------+------------------------------------------------------------+
|fsn-     |This "fs NOT" would be the logical "---" counterpart for the|
|         |fs.* prefixes above, butits similarity to fs will make      |
|         |reading difficult, so we stick with erbn-                   |
+---------+------------------------------------------------------------+
|All other|.. are also ---, and the only difference from erbn- is      |
|prefixes |aesthetical.                                                |
|         |                                                            |
+---------+------------------------------------------------------------+
|fsbot    | "Free software bot", an instance of fsbot.                 |
+---------+------------------------------------------------------------+
|         |                                                            |
|         |                                                            |
|         |                                                            |
|         |                                                            |
|         |                                                            |
+---------+------------------------------------------------------------+


ChangeLog
2007-12-06  D. Goel  <deego3@gmail.com>

	* erbc.el (fsi-replace-regexp): fix call to `error'.
	(fsi-merge-generic): Ditto.	
	(fsi-mv): "
	(fsi-replace-string): "
	and many other functions and files: Ditto.
	
2007-11-23  Dave Goel  <deego@gnufans.org>

	* COPYING: Replace by GPL v. 3

	* erbot.el and all other files: Change GPL v. 2 to 3.

2007-07-22  Michael Olson  <mwolson@gnu.org>

	* README.txt: Mention examples/dotemacs-mybot.

	* examples/dotemacs-mybot: New file that is the example .emacs for
	the bot.

2007-07-19  Michael Olson  <mwolson@gnu.org>

	* erbot.el (erbot-reply): Don't force the message through without
	flood protection.  Since ERC 5.1, ERC has very good flood
	protection, so make use of it.

2007-04-11  Vivek Dasmohapatra  <vivek@etla.org>

	* erbim.el (fs-unicode-describe): add usage instructions 
	(erbim-search-by-description): search for unicode characters by 
	description.
	(fs-unicode-find): bot-ui wrapper for erbim-search-by-description
	including usage message.
	(erbim-name-by-codepoint): changed output format to use #xXXX

2007-01-28  Vivek Dasmohapatra  <vivek@etla.org>

	* erbot.el (erbot-join-servers): `erc' takes :keyword style
	parameters in emacs22, the old argument list no longer works.
	Work out which erc version we're using and alter the call
	appropriately.

2007-01-27  Vivek Dasmohapatra  <vivek@etla.org>

	* erbutils.el (erbutils-describe-variable): help-button-xref
	has changed in emacs22, make this function work with it.

2006-11-28  Michael Olson  <mwolson@gnu.org>

	* erbot.el (erbot-join-servers): Make this work with the ERC
	development branch.

2006-09-28  D Goel  <deego@gnufans.org>

	* erbc.el (fsi-describe-from-english): smarter self search
	(fsi-generalize-search-term): new, for above.

2006-08-21  Michael Olson  <mwolson@gnu.org>

	* erbot.el (erbot-install): Remove check for erc-backend-version,
	since it no longer exists.  Use featurep instead.  This fixes a
	failure to join channels issue with the ERC development branch.

2006-05-18  Vivek Dasmohapatra  <vivek@etla.org>

	* erbtranslate.el (fsi-translate): If we can't make sense 
	of the call (not enough arguments) emit a usage message.

2006-05-17  Michael Olson  <mwolson@gnu.org>

	* contrib/translate.el (translate-load-pairs): Docfix.

	* erbtranslate.el (fsi-translate-list-services): Use
	translate-program, not erbn-translate-program, since the latter
	does not exist.

2006-05-12  Vivek Dasmohapatra  <vivek@etla.org>

	* contrib/translate.el: the symbol -> string coercion is not 
	required for arguments to translate.el, that's something
	specific to user-visible erbot functions.

	* erbtranslate.el: most translation code moved to translate.el
	fsi-translate-web-page temporarily disabled till I've tested
	it a bit more and made the implementation a little smarter
	than it currently is. (It should check ti see if a web page 
	service is actualy available)

2006-05-10  D Goel  <deego@gnufans.org>

	* erbtranslate.el: update authors

2006-05-10  Vivek Dasmohapatra  <vivek@etla.org>

	* erbtranslate.el (fsi-translate-list-pairs): destination language
	should also be searched for with a case insensitive predicate.

2006-05-09  D Goel  <deego@gnufans.org>

	* erbtranslate.el: update authors

2006-05-09  Vivek Dasmohapatra  <vivek@etla.org>

	* erbtranslate.el (fsi-translate-list-pairs): improved help, return
	more information when < N pairs, where N is currently 20. (Always 
	return full list if both origin and destination are both supplied)
	(erbtranslate-full-name): returns a "full name" for a language,
	consisting of all its human-readable aliases.
	(erbtranslate-unsupported-langs): list of languages emacs can't utf-8
	encode yet.
	(fsi-translate): abort for non-unicodable languages like Arabic and 
	Hebrew.
	(fsi-translate): Use full names of languages in error messages.
	(fsi-translate-list-pairs): Use full names of languages in messages.

2006-05-08  Vivek Dasmohapatra  <vivek@etla.org>

	* erbtranslate.el (erbtranslate-parse-pair): parse the output of
	erbn-translate-program --list-pairs.
	(erbtranslate-load-pairs): load the map of available translations.
	(fsi-translate-list-pairs): Alter this function so that it only lists
	possible translations when both an origin and a destination language
	have been specified. Otherwise just tells the user how many matching
	language pairs there are.
	(fsi-translate-list-pairs): coerce the arguments into strings first.
	(erbtranslate-parse-pair): some (one?) language codes are 3 letter.
	I thought this was non-canonical but that's what translate returns,
	so we must support it.
	(erbtranslate-parse-pair): tidy up error message
	(erbtranslate-req-to-pair): new. take the (possibly non-canonical)
	language names from the user and map them to the canonical language
	codes that translate expects.
	(fsi-translate): map human-friendly language names in translation 
	request to canonical language codes.
	(fsi-translate): use the coerced-strings, not the raw language args.

2006-05-08  D Goel  <deego@gnufans.org>

	* erbtranslate.el (erbn-translate-program): new.  Make the program customizeable.
	

	* contrib/shs.el: New shs.el, cleans up temporary files.

2006-04-24  D Goel  <deego@gnufans.org>

	* erbtranslate.el: DECLARE SAFE. Uses call-process now.

	* erball.el: require erbtranslate.

	* erbutils.el (erbutils-enabled-check): new function.  

	* erbtranslate.el: revamp the file to make it secure, not yet final.

2006-04-20  D Goel  <deego@gnufans.org>

	* erbot.el (erbot-safe-p): Improve this function a bit.
	(erbot-safe-nocontrol-p): new.

	* erbot-lispy.el (erbot-lispy-safe-p): change name from erbot-lispy-safep

	* erbot.el (erbot-safe-p): change name from erbot-safep

	* erbc.el (erbn-url-functions-p): new variable.  Disallow url's
	unless enabled here.  Else potential freeze.
	(erbn-internal-web-page-time): rename from the one below.
	(fs-internal-web-page-time): rename to the one above

erball.el
;;; erball.el --- Functions on all files. 
;; Time-stamp: <2006-04-24 13:43:38 deego>
;; Copyright (C) 2002 D. Goel
;; Emacs Lisp Archive entry
;; Filename: erbc.el
;; Package: erbc
;; Author: D. Goel <deego@gnufans.org>
;; Version: 



;; Usually maintenance
;; not all of these may be required depending on how you use erbot..
(require 'cl)

;; Compilation

(defvar erball-compilation-paths-rel-to
  (let (args ret)
    (while command-line-args-left
      (if (string= "--paths-rel-to" (car command-line-args-left))
          (progn
            (setq ret (cadr command-line-args-left))
            (setq command-line-args-left (cddr command-line-args-left)))
        (add-to-list 'args (car command-line-args-left) t)
        (setq command-line-args-left (cdr command-line-args-left))))
    (setq command-line-args-left args)
    ret)
  "Text to be prepended to each element in `erball-compilation-paths'.
Can be specified by passing \"--paths-rel-to ARG\" on the emacs
command line.
This value is also added to the load-path.
A trailing backslash is required.")

(defun erball-assoc-string (key list)
  "Like `assoc' but specifically for strings."
  (if (fboundp 'assoc-string)
      (assoc-string key list)
    (catch 'found
      (dolist (el list)
        (when (string= key el)
          (throw 'found el))))))

(defvar erball-compiling-p
  (if (erball-assoc-string "--compile-erbot" command-line-args-left)
      (progn
        (message "%s" (concat "\nCompiling source in "
			      (file-name-nondirectory (expand-file-name "."))
			      " ...\n"))
        (setq command-line-args-left
              (delete "--compile-erbot" command-line-args-left))
        t)
    nil)
  "Determine whether erbot is currently being compiled.")

(defcustom erball-compilation-paths
  '("contrib"
    ".."
    "../erc"
    "../bbdb/lisp")
  "Elements to add to the load path during compilation.
If `erball-compilation-paths-rel-to' is specified, it is
prepended to each element and also added verbatim to the path.
The current directory is automatically added to the path."
  :group 'erball
  )

(when erball-compiling-p
  (add-to-list 'load-path ".")
  (when erball-compilation-paths-rel-to
    (add-to-list 'load-path erball-compilation-paths-rel-to))
  (dolist (dir erball-compilation-paths)
    (add-to-list 'load-path
                 (concat erball-compilation-paths-rel-to dir))))

;; Load all erbot files

(defmacro erball-ignore-errors-loudly (&rest body)
  "Like ignore-errors, but tells the error..

Copied from deego's `ignore-errors-my', which owes some of its work
to: Kalle on 7/3/01:
 * used backquote: something i was too lazy to convert my macro to..
 * removed the progn: condition-case automatically has one..
 * made sure that the return is nil.. just as it is in ignore-errors. "
  (let ((err (gensym)))
    `(condition-case ,err (progn ,@body)
       (error
	(ding t)
	(ding t)
	(ding t)
	(message "IGNORED ERROR: %s" (error-message-string ,err))
	(sit-for 1)
	nil))))



(erball-ignore-errors-loudly (require 'bbdb))
(erball-ignore-errors-loudly (require 'doctor))
(erball-ignore-errors-loudly (require 'erc))
(erball-ignore-errors-loudly (require 'erc-stamp))



(unless noninteractive (erball-ignore-errors-loudly (require 'dunnet)))
(erball-ignore-errors-loudly (require 'erbot))
(erball-ignore-errors-loudly (require 'erbcountry))
(erball-ignore-errors-loudly (require 'erbutils))
(erball-ignore-errors-loudly (require 'erblog))
(erball-ignore-errors-loudly (require 'erbeng))
(erball-ignore-errors-loudly (require 'erbdata))
(erball-ignore-errors-loudly (require 'erbkarma))
(erball-ignore-errors-loudly (require 'erblisp))
(erball-ignore-errors-loudly (require 'erbc))
(erball-ignore-errors-loudly (require 'erbc2))
(erball-ignore-errors-loudly (require 'erbc3))
(erball-ignore-errors-loudly (require 'erbc4))
(erball-ignore-errors-loudly (require 'erbc5))
(erball-ignore-errors-loudly (require 'erbc6))
(erball-ignore-errors-loudly (require 'erbcspecial))
(erball-ignore-errors-loudly (require 'erbbdb))
(erball-ignore-errors-loudly (require 'erbforget))
(erball-ignore-errors-loudly (require 'erbedit))
(erball-ignore-errors-loudly (require 'erbtrain))
(erball-ignore-errors-loudly (require 'erbwiki))
(erball-ignore-errors-loudly (require 'erbunlisp))
(erball-ignore-errors-loudly (require 'erbcompat))

(erball-ignore-errors-loudly (require 'erbmsg))
(erball-ignore-errors-loudly (require 'erbtranslate))
(erball-ignore-errors-loudly (require 'erbim))

(erball-ignore-errors-loudly (require 'flame))

(erball-ignore-errors-loudly (require 'mkback))
(erball-ignore-errors-loudly (require 'lines))
(erball-ignore-errors-loudly (require 'google))
(erball-ignore-errors-loudly (require 'oct))

;; the rest of the commands here are useful to the author when editing erbot. 

(defcustom erball-files
  (if erball-compiling-p
      (directory-files "." nil "\.el$")
    '("erbot.el"
      "erbutils.el"
      "erblog.el"
      "erbeng.el"
      "erbcountry.el"
      "erbdata.el"
      "erbedit.el"
      "erbforget.el"
      "erbkarma.el"
      "erblisp.el"
      "erbunlisp.el"
      "erbtrain.el"
      "erbwiki.el"
      "erbc.el"
      "erbc2.el"
      "erbc3.el"
      "erbc4.el"
      "erbc5.el"
      "erbc6.el"
      ))

    ""
    :group 'erball
    )

(defun erball-reload ()
  (interactive)
  (mapcar
   'load
   erball-files))

(defun erball-visit ()
  (interactive)
  (mapcar
   (lambda (a)
     (find-file (locate-library a))
     (auto-revert-mode 1))
   erball-files))

;;;###autoload
(defun erball-compile ()
  (interactive)
  (if erball-compiling-p
      (progn
        (ignore-errors (erball-reload))
        (mapcar
         (lambda (arg)
           (erball-ignore-errors-loudly
            (byte-compile-file arg)))
         erball-files)
        (message "\nCompilation complete!\n"))
    (ignore-errors (kill-buffer "*Compile-Log*"))
    (erball-visit)
    (erball-reload)
    (mapcar
     (lambda (arg)
       (erball-ignore-errors-loudly
        (byte-compile-file (locate-library arg))))
     erball-files)
    (switch-to-buffer "*Compile-Log*")
    (delete-other-windows)
    (goto-char (point-min))))


(provide 'erball)
